File: templates/tree/bst.go
package tree

func IsValidBST(root *TreeNode) bool {
	var dfs func(*TreeNode, *int, *int) bool
	dfs = func(node *TreeNode, low, high *int) bool {
		if node == nil {
			return true
		}
		if low != nil && node.Val <= *low {
			return false
		}
		if high != nil && node.Val >= *high {
			return false
		}
		return dfs(node.Left, low, &node.Val) && dfs(node.Right, &node.Val, high)
	}
	return dfs(root, nil, nil)
}

func KthSmallest(root *TreeNode, k int) int {
	st := []*TreeNode{}
	cur := root
	for cur != nil || len(st) > 0 {
		for cur != nil {
			st = append(st, cur)
			cur = cur.Left
		}
		cur = st[len(st)-1]
		st = st[:len(st)-1]
		k--
		if k == 0 {
			return cur.Val
		}
		cur = cur.Right
	}
	return -1
}

File: templates/tree/bst.py
from .tree import TreeNode


def is_valid_bst(root: TreeNode | None) -> bool:
    def dfs(node: TreeNode | None, low: int | None, high: int | None) -> bool:
        if node is None:
            return True
        if low is not None and node.val <= low:
            return False
        if high is not None and node.val >= high:
            return False
        return dfs(node.left, low, node.val) and dfs(node.right, node.val, high)

    return dfs(root, None, None)


def kth_smallest(root: TreeNode | None, k: int) -> int:
    stack: list[TreeNode] = []
    current = root

    while current is not None or stack:
        while current is not None:
            stack.append(current)
            current = current.left
        current = stack.pop()
        k -= 1
        if k == 0:
            return current.val
        current = current.right

    return -1
